jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java

```diff
@@ -284,27 +285,36 @@ public void closedBySendError()
         }
     }
 
-    /**
-     * Make {@link #close()} am asynchronous method
-     * @param closeCallback The callback to use when close() is called.
-     */
-    public void setClosedCallback(Callback closeCallback)
+    public void close(Closeable wrapper, Callback callback)
     {
-        _closeCallback = closeCallback;
-    }
-
-    public Callback getClosedCallback()
-    {
-        return _closeCallback;
+        _closeCallback = callback;
+        try
+        {
+            if (wrapper != null)
+                wrapper.close();
+            if (!isClosed())
+                close();
+        }
+        catch (Throwable th)
+        {
+            closed();
+            if (_closeCallback == null)
+                LOG.ignore(th);
+            else
+                callback.failed(th);
+        }
+        finally
+        {
+            if (_closeCallback != null)
+                callback.succeeded();
+            _closeCallback = null;
+        }
     }
 
     @Override
     public void close()
     {
-        Callback closeCallback = _closeCallback;
-        _closeCallback = null;
-        if (closeCallback == null)
-            closeCallback = BLOCKING_CLOSE_CALLBACK;
+        Callback closeCallback = _closeCallback == null ? BLOCKING_CLOSE_CALLBACK : _closeCallback;
```

Comment:
Uhm. Now that you have a `close(Callback)` method, this method can be implemented in terms of the other.

```java
public void close()
{
    FutureCallback callback = new FutureCallback();
    close(callback);
    callback.await();
}
```
The other method does all the lifting.

My point being that we should avoid having the `_closeCallback` field and continuously test for it being non-null and completing it. It should just be a local variable in the blocking version of the `close()` method.
}

Reply:
@sbordet sorry but it can't be done this way.   The new close method is passed a closeable, which either be null or the Writer used for the output.  The writer is closed first and if all goes OK, it will be the writer that calls HttpOutput.close(), so there is no possibility to pass the callback through the writer.    Thus a field is really the simplest way to capture this.